udtube/callbacks.py

```diff
@@ -49,38 +49,47 @@ def write_on_batch_end(
         batch_idx: int,
         dataloader_idx: int,
     ) -> None:
-        y_upos_hat = (
+        # Batch-level argmax on the classification heads.
+        upos_hat = (
             torch.argmax(logits.upos, dim=1) if logits.use_upos else None
         )
-        y_xpos_hat = (
+        xpos_hat = (
             torch.argmax(logits.xpos, dim=1) if logits.use_xpos else None
         )
-        y_lemma_hat = (
+        lemma_hat = (
             torch.argmax(logits.lemma, dim=1) if logits.use_lemma else None
         )
-        y_feats_hat = (
+        feats_hat = (
             torch.argmax(logits.feats, dim=1) if logits.use_feats else None
         )
         for i, tokenlist in enumerate(batch.tokenlists):
-            # Edits the relevant fields in each tokenlist.
-            if y_upos_hat is not None:
-                upos_hat = self.mapper.decode_upos(y_upos_hat[i, :])
-                for j, upos in enumerate(upos_hat[: len(tokenlist)]):
-                    tokenlist[j]["upos"] = upos
-            if y_xpos_hat is not None:
-                xpos_hat = self.mapper.decode_xpos(y_xpos_hat[i, :])
-                for j, xpos in enumerate(xpos_hat[: len(tokenlist)]):
-                    tokenlist[j]["xpos"] = xpos
-            if y_lemma_hat is not None:
-                lemma_hat = self.mapper.decode_lemma(
-                    [token["form"] for token in tokenlist],
-                    y_lemma_hat[i, :],
+            # Sentence-level decoding of the classification indices, followed
+            # by rewriting the fields in the tokenlist.
+            if upos_hat is not None:
+                upos_it = self.mapper.decode_upos(upos_hat[i, :])
+                for token in tokenlist:
+                    if token.is_mwe:
+                        continue
```

Comment:
I am certainly missing something here, but can you explain why there's no `_ = next(upos_it)` here to skip over the item? Maybe with a little more review it will reveal itself to me, but leaving this comment (same for 78)

Reply:
In case of MWEs tokenlist is longer than the set of tags, so not needed. (UD convention is that there are no morph tags for MWEs I believe.) I'll add a comment.